### PR TITLE
feat: close transports that implement io.Closer

### LINKED ISSUE #999: Proposal: Closable Transport
# Use Case
While writing a tor transport I noticed that some transports runs some node beyond the listener (mainly overlay networks).
For example my tor transport runs a tor node and I need to close it at one point. I can't rely on the listeners to close it because it's not because a node stop listening on an address it will not then dial someone.
# Additions
I propose to add this in `go-libp2p-core/transport` :
```go
type ClosableTransport interface {
  Transport

  // Close is called when the libp2p node doesn't need the transport anymore (the is node closing).
  // After that the transport can safely assume no more Listen() or Dial() are gonna be done.
  // This is usefull for transport running overlay network.
  Close() error
}
```
Then in `host.Close()` the host would use reflection to terminate the transport when closing (after closing all listeners) :
```go
// Not actual working code, just an idea
var wg sync.WaitGroup
for _, v := range h.transports {
  ct, ok := v.(tpt.ClosableTransport)
  if ok {
    wg.Add(1)
    go func (t tpt.ClosableTransport){ // Needed because range is not thread safe
      defer wg.Done()
      t.Close()
    }(ct)
  }
}
wg.Wait()
```
(Also I can't rely on the garbage collector to do this work because I have some CGO code (or a system fork) and the garbage collector doesn't deal with that.)
## In the Future
If in the future a method to unhook transport from a libp2p is added the transport will be closable that way.

### PR DESCRIPTION
**Pull request by Stebalien**
This way, transports with shared resources (e.g., reused sockets) can clean them up. Otherwise, we have no way to tell, e.g., the QUIC transport to close it's socket "pool".

fixes https://github.com/libp2p/go-libp2p/issues/999

### CODE REVIEW

**Review by vyzo — APPROVED** [review]

File: swarm.go
Comment (Jorropo) [inline swarm.go:188]: In my case (tor and i2p) `closer.Close()` is slow, it needs to call other peers in the dht about him leaving, that takes some times, I would like each transport to have his own goroutine.

**Review by Jorropo — CHANGES_REQUESTED** [review]

**Review by Jorropo — APPROVED** [review]

**Review by Jorropo — APPROVED** [review]

File: swarm.go
Comment (Stebalien) [reply swarm.go]: Eh, I'd prefer to be able to _wait_ and I'm not too concerned about the cost of firing off goroutines on shutdown.

**Review by vyzo — APPROVED** [review]

**Review by Jorropo — APPROVED** [review]
LGTM I couldn't ask for more 👌

**Review by Jorropo — CHANGES_REQUESTED** [review]

File: swarm.go
Comment (Stebalien) [reply swarm.go:187]: That isn't safe if `Close` is called concurrently. Basically, we can only call `Add` on this waitgroup when we can guarantee that the waitgoup won't be zero.

File: swarm.go
Comment (Stebalien) [reply swarm.go:180]: I'm not going to take a lock in one function, and release it in another. The correct fix here would be to call `refs.Add(1)` *before* starting a dial, releasing the ref on dial failure.

**Comment by jacobheun** [comment]
What's the state of this, is it blocked by anything?

**Comment by vyzo** [comment]
not that I know, LGTM. @Stebalien ?

**Comment by Stebalien** [comment]
I haven't merged it because it's still technically possible to call `Dial` on transports after calling `Close`. Honestly, merging with that issue is probably fine for now.

**Review by Jorropo — APPROVED** [review]
I've changed my mind, that a close, as long as it works performance is not an issue.

**Comment by Jorropo** [comment]
@Stebalien what is the status of this ?
I need it for https://github.com/berty/go-libp2p-tor-transport, without it this is causing memory, cpu and process leakage.
Plus I think there is no need to bother as much as I were doing, it's fine for me.

**Comment by Stebalien** [comment]
Yeah, this is probably fine as-is. I'll merge for now with an open issue: https://github.com/libp2p/go-libp2p-swarm/issues/230.

**Comment by Stebalien** [comment]
Actually, I do need a final signoff from @raulk on this.

File: swarm.go
Comment (raulk) [inline swarm.go:193]: A log statement confirming that we're closing down this closeable transport would be nice. But then again, we need better logging all throughout, so I won't object if we don't add this one here right now.

File: swarm_listen.go
Comment (raulk) [inline swarm_listen.go:45]: I'm not sure I understand why this here? Which edge case is this addressing? Either we accept or we don't accept adding listen addresses while the swarm is closing/closed, more generally, but this looks weird?

File: swarm_transport.go
Comment (raulk) [inline swarm_transport.go:55]: Would be nice to actually model a swarm status with atomic ints. Inferring we're closing down based on nil vs. 0-length slice differentiation is a bit of a footgun.

**Review by raulk — COMMENTED** [review]
I'm ok to merge these changes, but there are some edge cases that I quite don't understand. And the swarm needs to store its state (unstarted, started, closing, closed) in an atomic int so we can implement a state machine, rather than rely on hacks to infer which state we're in.

File: swarm_transport.go
Comment (Stebalien) [reply swarm_transport.go:55]: Note: this is a map, not a slice. We could have a second flag, but I'd prefer to avoid storing the state twice.

File: swarm_listen.go
Comment (Stebalien) [reply swarm_listen.go:45]: I'll add a comment: TL;DR: we could fail to get the transport either because we're closed, or because the transport was never registered.

**Comment by Stebalien** [comment]
Tests are broken due to #246. Otherwise, they passed.

**Pull request merged**
